src/bitops.c

```diff
@@ -697,59 +894,94 @@ void bitopCommand(client *c) {
     /* Compute the bit operation, if at least one string is not empty. */
     if (maxlen) {
         res = (unsigned char*) sdsnewlen(NULL,maxlen);
-        unsigned char output, byte;
+        unsigned char output, byte, disjunction, common_bits;
         unsigned long i;
+        int useAVX2 = 0;
+
+        /* Number of bytes processed from each source key */
+        j = 0;
+
+#if defined(HAVE_AVX2)
+        if (BITOP_USE_AVX2) {
+            j = bitopCommandAVX(src, res, op, numkeys, minlen);
 
-        /* Fast path: as far as we have data for all the input bitmaps we
+            serverAssert(minlen >= j);
+            minlen -= j;
+
+            useAVX2 = 1;
+        }
+#endif
+
+#if !defined(USE_ALIGNED_ACCESS)
+        /* We don't have AVX2 but we still have fast path:
+         * as far as we have data for all the input bitmaps we
          * can take a fast path that performs much better than the
          * vanilla algorithm. On ARM we skip the fast path since it will
          * result in GCC compiling the code using multiple-words load/store
          * operations that are not supported even in ARM >= v6. */
-        j = 0;
-        #ifndef USE_ALIGNED_ACCESS
-        if (minlen >= sizeof(unsigned long)*4 && numkeys <= 16) {
-            unsigned long *lp[16];
+        if (minlen >= sizeof(unsigned long)*4) {
+            /* We can't have entered the AVX2 path since minlen >= sizeof(unsigned long)*4
+             * AVX2 path operates on steps of sizeof(__m256i) which for 64-bit
+             * machines (the only ones supporting AVX2) is equal to
+             * sizeof(unsigned long)*4. That means after the AVX2
+             * path minlen will necessarily be < sizeof(unsigned long)*4. */
+            serverAssert(!useAVX2);
+
+            unsigned long **lp = (unsigned long**)src;
             unsigned long *lres = (unsigned long*) res;
+            unsigned long *first_key = (unsigned long*)src[0];
+            unsigned long lcommon_bits[4];
+
+            const size_t step = sizeof(unsigned long)*4;
+            size_t processed = 0;
```

Comment:
move this to `else if (op == BITOP_DIFF || op == BITOP_DIFF1 || op == BITOP_ANDOR) {`?